### PR TITLE
lib: Improve password fields

### PR DESCRIPTION
**Pull request by skobyda**
- Adds show/hide feature
- Adds a minimum to a progress bar
- Changes width of a progress bar
- Shows message indicating the strength of the password

Fixes https://github.com/cockpit-project/cockpit/issues/19199

[Screencast from 2023-08-22 14-30-21.webm](https://github.com/cockpit-project/cockpit/assets/42733240/231f89ec-04b4-474e-adb1-95b405f5e778)

### CODE REVIEW

**Comment by skobyda** [comment]
Currently we these 4 levels of messages for passwords "Weak password" < "Good password" < "Strong password" < "Excellent password".
I'm not sure about the wording of the "Good password". It has score more than 33 but less than 66 (and is shown in yellow color). It's a password which is neither good nor bad.
Maybe it should be "acceptable password"? 
WDTY @garrett

File: pkg/lib/cockpit-components-password.jsx [hunk 1]
```diff
@@ -106,19 +127,35 @@ export const PasswordFormFields = ({
...
-                              title={_("password quality")}
-                              size={ProgressSize.sm}
-                              measureLocation={ProgressMeasureLocation.none}
-                              variant={variant}
-                              value={Number.isInteger(passwordStrength) ? passwordStrength : 0} />
-                    <div id={idPrefix + "-password-meter-message"} className="pf-v5-c-form__helper-text" aria-live="polite">{passwordMessage}</div>
-                </div>
+                <InputGroup>
+                    <InputGroupItem isFill>
+                        <TextInput className="check-passwords" type={passwordHidden ? "password" : "text"} id={idPrefix + "-pw1"}
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-password.jsx:132]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20230822-125449-71e7c813-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/lib/cockpit-components-password.jsx [hunk 2]
```diff
@@ -106,19 +127,35 @@ export const PasswordFormFields = ({
...
+                                   autoComplete="new-password" value={password} onChange={(_event, value) => onPasswordChanged(value)}
+                                   validated={error_password ? "warning" : "default"} />
+                    </InputGroupItem>
+                    <InputGroupItem>
+                        <Button
+                            variant="control"
+                            onClick={() => setPasswordHidden(!passwordHidden)}
+                            aria-label={ passwordHidden ? _("Show password") : _("Hide password")}>
+                            {passwordHidden ? <EyeIcon /> : <EyeSlashIcon />}
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-password.jsx:141]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20230822-125449-71e7c813-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/lib/cockpit-components-password.jsx [hunk 3]
```diff
@@ -131,8 +168,20 @@ export const PasswordFormFields = ({
             {password_confirm_label && <FormGroup label={password_confirm_label}
                        id={idPrefix + "-pw2-group"}
                        fieldId={idPrefix + "-pw2"}>
-                <TextInput type="password" id={idPrefix + "-pw2"} autoComplete="new-password"
-                           value={passwordConfirm} onChange={(_event, value) => { setConfirmPassword(value); change("password_confirm", value) }} />
+                <InputGroup>
+                    <InputGroupItem isFill>
+                        <TextInput type={passwordConfirmHidden ? "password" : "text"} id={idPrefix + "-pw2"} autoComplete="new-password"
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-password.jsx:173]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20230822-125449-71e7c813-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/lib/cockpit-components-password.jsx [hunk 4]
```diff
@@ -131,8 +168,20 @@ export const PasswordFormFields = ({
...
+                            value={passwordConfirm} onChange={(_event, value) => { setConfirmPassword(value); change("password_confirm", value) }} />
+                    </InputGroupItem>
+                    <InputGroupItem>
+                        <Button
+                            variant="control"
+                            onClick={() => setPasswordConfirmHidden(!passwordConfirmHidden)}
+                            aria-label={passwordConfirmHidden ? _("Show confirmation password") : _("Hide confirmation password")}>
+                            {passwordConfirmHidden ? <EyeIcon /> : <EyeSlashIcon />}
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-password.jsx:181]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20230822-125449-71e7c813-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by martinpitt** [comment]
needsdesign → moving to draft

**Comment by garrett** [comment]
image

The contrast between the text and its background is not high enough. The color of the text should be the darker gold.

You can see this in PF demos:

image
https://www.patternfly.org/components/helper-text/html-demos/#helper-text-on-a-form

Here, you can see how PF uses brighter versions for icons, progress bars, etc. and the darker related color for text:
 
image
https://www.patternfly.org/components/alert/html/#alert-examples

Colors are specified here:

image
https://www.patternfly.org/design-foundations/colors/#primary-colors

- icon: `--pf-v5-global--warning-color--100`
- text: `--pf-v5-global--warning-color--200`

This isn't just a problem with the warning colors either, but it's most noticeable with it, as it has the least contrast. The other colors need to use the text variants too.

**Comment by garrett** [comment]
There's a strange intermediate step where the strength meter is shown in a blue state without text, right at the beginning of typing:

image

**Comment by garrett** [comment]
> I'm not sure about the wording of the "Good password". It has score more than 33 but less than 66 (and is shown in yellow color). It's a password which is neither good nor bad.
> Maybe it should be "acceptable password"?

Yes. :+1:

**Comment by skobyda** [comment]
Updated: Changed to the text color, got rid of the blue progress bar without any text and from now on using "Acceptable" instead of "Good" password:

Screenshot from 2023-09-18 14-02-44

File: pkg/lib/cockpit-components-password.jsx [hunk 5]
```diff
@@ -77,20 +82,34 @@ export const PasswordFormFields = ({
                         setPasswordMessage(strength.message);
                     });
         } else {
-            setPasswordStrength("");
+            setPasswordStrength();
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-password.jsx:85]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20230918-120248-8fd6ef98-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/lib/cockpit-components-password.jsx [hunk 6]
```diff
@@ -106,19 +125,35 @@
...
-                              title={_("password quality")}
-                              size={ProgressSize.sm}
-                              measureLocation={ProgressMeasureLocation.none}
-                              variant={variant}
-                              value={Number.isInteger(passwordStrength) ? passwordStrength : 0} />
-                    <div id={idPrefix + "-password-meter-message"} className="pf-v5-c-form__helper-text" aria-live="polite">{passwordMessage}</div>
-                </div>
+                <InputGroup>
+                    <InputGroupItem isFill>
+                        <TextInput className="check-passwords" type={passwordHidden ? "password" : "text"} id={idPrefix + "-pw1"}
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-password.jsx:130]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20230918-120248-8fd6ef98-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/lib/cockpit-components-password.jsx [hunk 7]
```diff
@@ -106,19 +125,35 @@
...
+                                   autoComplete="new-password" value={password} onChange={(_event, value) => onPasswordChanged(value)}
+                                   validated={error_password ? "warning" : "default"} />
+                    </InputGroupItem>
+                    <InputGroupItem>
+                        <Button
+                            variant="control"
+                            onClick={() => setPasswordHidden(!passwordHidden)}
+                            aria-label={ passwordHidden ? _("Show password") : _("Hide password")}>
+                            {passwordHidden ? <EyeIcon /> : <EyeSlashIcon />}
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-password.jsx:139]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20230918-120248-8fd6ef98-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/lib/cockpit-components-password.jsx [hunk 8]
```diff
@@ -131,8 +166,20 @@
             {password_confirm_label && <FormGroup label={password_confirm_label}
                        id={idPrefix + "-pw2-group"}
                        fieldId={idPrefix + "-pw2"}>
-                <TextInput type="password" id={idPrefix + "-pw2"} autoComplete="new-password"
-                           value={passwordConfirm} onChange={(_event, value) => { setConfirmPassword(value); change("password_confirm", value) }} />
+                <InputGroup>
+                    <InputGroupItem isFill>
+                        <TextInput type={passwordConfirmHidden ? "password" : "text"} id={idPrefix + "-pw2"} autoComplete="new-password"
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-password.jsx:171]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20230918-120248-8fd6ef98-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/lib/cockpit-components-password.jsx [hunk 9]
```diff
@@ -131,8 +166,20 @@
...
+                            value={passwordConfirm} onChange={(_event, value) => { setConfirmPassword(value); change("password_confirm", value) }} />
+                    </InputGroupItem>
+                    <InputGroupItem>
+                        <Button
+                            variant="control"
+                            onClick={() => setPasswordConfirmHidden(!passwordConfirmHidden)}
+                            aria-label={passwordConfirmHidden ? _("Show confirmation password") : _("Hide confirmation password")}>
+                            {passwordConfirmHidden ? <EyeIcon /> : <EyeSlashIcon />}
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-password.jsx:179]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20230918-120248-8fd6ef98-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Review by garrett — CHANGES_REQUESTED** [review]
It's a lot better. Thanks!

However, the meter (progress bar) doesn't update based on the strength.

[Screencast from 2023-09-20 09-59-43.webm](https://github.com/cockpit-project/cockpit/assets/10246/1bf48a36-7f82-498c-9386-2d601b34bef3)

File: pkg/lib/cockpit-components-password.jsx (hunk 5 above)
Comment (cockpituous) [inline pkg/lib/cockpit-components-password.jsx:85]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20231016-125913-03c3b351-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/lib/cockpit-components-password.jsx [hunk 10]
```diff
@@ -106,19 +125,35 @@ export const PasswordFormFields = ({
...
-                              title={_("password quality")}
-                              size={ProgressSize.sm}
-                              measureLocation={ProgressMeasureLocation.none}
-                              variant={variant}
-                              value={Number.isInteger(passwordStrength) ? passwordStrength : 0} />
-                    <div id={idPrefix + "-password-meter-message"} className="pf-v5-c-form__helper-text" aria-live="polite">{passwordMessage}</div>
-                </div>
+                <InputGroup>
+                    <InputGroupItem isFill>
+                        <TextInput className="check-passwords" type={passwordHidden ? "password" : "text"} id={idPrefix + "-pw1"}
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-password.jsx:130]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20231016-125913-03c3b351-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/lib/cockpit-components-password.jsx [hunk 11]
```diff
@@ -106,19 +125,35 @@ export const PasswordFormFields = ({
...
+                                   autoComplete="new-password" value={password} onChange={(_event, value) => onPasswordChanged(value)}
+                                   validated={error_password ? "warning" : "default"} />
+                    </InputGroupItem>
+                    <InputGroupItem>
+                        <Button
+                            variant="control"
+                            onClick={() => setPasswordHidden(!passwordHidden)}
+                            aria-label={ passwordHidden ? _("Show password") : _("Hide password")}>
+                            {passwordHidden ? <EyeIcon /> : <EyeSlashIcon />}
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-password.jsx:139]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20231016-125913-03c3b351-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/lib/cockpit-components-password.jsx [hunk 12]
```diff
@@ -131,8 +166,20 @@ export const PasswordFormFields = ({
             {password_confirm_label && <FormGroup label={password_confirm_label}
                        id={idPrefix + "-pw2-group"}
                        fieldId={idPrefix + "-pw2"}>
-                <TextInput type="password" id={idPrefix + "-pw2"} autoComplete="new-password"
-                           value={passwordConfirm} onChange={(_event, value) => { setConfirmPassword(value); change("password_confirm", value) }} />
+                <InputGroup>
+                    <InputGroupItem isFill>
+                        <TextInput type={passwordConfirmHidden ? "password" : "text"} id={idPrefix + "-pw2"} autoComplete="new-password"
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-password.jsx:171]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20231016-125913-03c3b351-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/lib/cockpit-components-password.jsx [hunk 13]
```diff
@@ -131,8 +166,20 @@ export const PasswordFormFields = ({
...
+                            value={passwordConfirm} onChange={(_event, value) => { setConfirmPassword(value); change("password_confirm", value) }} />
+                    </InputGroupItem>
+                    <InputGroupItem>
+                        <Button
+                            variant="control"
+                            onClick={() => setPasswordConfirmHidden(!passwordConfirmHidden)}
+                            aria-label={passwordConfirmHidden ? _("Show confirmation password") : _("Hide confirmation password")}>
+                            {passwordConfirmHidden ? <EyeIcon /> : <EyeSlashIcon />}
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-password.jsx:179]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20231016-125913-03c3b351-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by skobyda** [comment]
Fixed the issue @garrett 

[Screencast from 2023-10-17 15-06-03.webm](https://github.com/cockpit-project/cockpit/assets/42733240/9c50c190-0faa-4a2c-b1d5-cc4f563aaa8b)

File: pkg/lib/cockpit-components-password.scss [hunk 14]
```diff
@@ -1,7 +1,12 @@
...
+
 .ct-password-strength-meter {
   grid-gap: var(--pf-v5-global--spacer--xs);
 
   .pf-v5-c-progress__description, .pf-v5-c-progress__status {
     display: none;
   }
+
+  inline-size: var(--pf-v5-global--spacer--2xl);
 }
```
Comment (garrett) [inline pkg/lib/cockpit-components-password.scss:12]:
```suggestion
.ct-password-strength-meter {
  grid-gap: var(--pf-v5-global--spacer--xs);
  inline-size: var(--pf-v5-global--spacer--2xl);

  .pf-v5-c-progress__description, .pf-v5-c-progress__status {
    display: none;
  }
}
```

Parent CSS should be grouped together at the top. (This doesn't change anything, but it's better when it's not scattered.)

**Review by garrett — APPROVED** [review]
This was waiting on the password meter fix. Thanks! Approved.

(But please move that line of CSS.)

File: pkg/lib/cockpit-components-password.jsx (hunk 13 above)
Comment (cockpituous) [inline pkg/lib/cockpit-components-password.jsx:179]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20231030-135959-d54860fd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/lib/cockpit-components-password.jsx (hunk 5 above)
Comment (cockpituous) [inline pkg/lib/cockpit-components-password.jsx:85]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20231031-011601-1bae20a1-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/lib/cockpit-components-password.jsx [hunk 15]
```diff
@@ -120,19 +122,19 @@ export const PasswordFormFields = ({
                         <Button
                             variant="control"
                             onClick={() => setPasswordHidden(!passwordHidden)}
-                            aria-label={passwordHidden ? _("Show password") : _("Hide password")}>
+                            aria-label={ passwordHidden ? _("Show password") : _("Hide password")}>
```
Comment (martinpitt) [inline pkg/lib/cockpit-components-password.jsx]: This is inconsistent spacing, and it also doesn't belong into the "lib: Change width of password progress" commit. This bit should be reverted.

File: pkg/lib/cockpit-components-password.jsx [hunk 16]
```diff
@@ -120,19 +122,19 @@ export const PasswordFormFields = ({
...
                             {passwordHidden ? <EyeIcon /> : <EyeSlashIcon />}
                         </Button>
                     </InputGroupItem>
                 </InputGroup>
                 <div>
                     <Progress id={idPrefix + "-meter"}
-                              className={"ct-password-strength-meter " + variant}
```
Comment (martinpitt) [inline pkg/lib/cockpit-components-password.jsx:114]: This previous indentation is better to read IMHO.

**Review by martinpitt — CHANGES_REQUESTED** [review]
Thanks! The code changes look good, but the committed pixel changes [look](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20231031-011601-1bae20a1-fedora-38-storage/log.html) [like](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20231031-011601-1bae20a1-fedora-38-other/log.html) they revert #19520. I'll rebase this PR, fix the above two little issues, and re-do the pixel updates.

File: pkg/lib/cockpit-components-password.jsx (hunk 16 above)
Comment (martinpitt) [reply pkg/lib/cockpit-components-password.jsx:114]: Ah, other code in that file uses 4-space indentation as well, so ok.

**Comment by martinpitt** [comment]
That looks much saner -- most tests are okay, just the [accounts tests](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20231031-045839-c175744c-fedora-38-other/log.html) have the updated pixels for the reveal buttons. I'll push these.

**Review by martinpitt — APPROVED** [review]
Thanks!

File: pkg/lib/cockpit-components-password.jsx (hunk 5 above)
Comment (cockpituous) [inline pkg/lib/cockpit-components-password.jsx:85]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20231031-052537-1ade659f-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Pull request merged**